src/render/tutorials/graphql-usage/2_0.html.md

```diff
@@ -0,0 +1,777 @@
+
+## Step 2
+We would like to fetch data from 3rd parties using existing Knot.x mechanisms, not to reinvent the wheel. Ideally we would have a generic data fetcher that can process Knot.x tasks defined in configuration. 
+
+Our fetcher will parse configuration to create tasks and run them using `FragmentsEngine`. It will output a `Fragment` that will contain data fetched from google in it's payload. Fragment's payload is just a `JsonObject` where actions can store data.
+
+
+#### Execute tasks from configuration
+Let's start with something like this:
+
+```java
+public abstract class TaskDataFetcher<T> implements DataFetcher<CompletableFuture<T>> {
+
+    private final Vertx vertx;
+    private final JsonObject config;
+    private final RoutingContext routingContext;
+    private final String taskName;
+    private final FragmentsEngine engine;
+
+    TaskDataFetcher(VertX, JsonObject config, RoutingContext routingContext, String taskName) {
+        this.vertx = vertx;
+        this.config = config;
+        this.routingContext = routingContext;
+        this.taskName = taskName;
+        engine = new FragmentsEngine(vertx);
+    }
+
+    @Override
+    public CompletableFuture<T> get(DataFetchingEnvironment environment) {
+        FragmentEventContextTaskAware eventContextTaskAware = setupTask(vertx, config, routingContext, environment);
+        CompletableFuture<T> future = new CompletableFuture<>();
+
+        engine
+            .execute(Collections.singletonList(eventContextTaskAware))
+            .subscribe(events -> {
+                // Here we'll get the outcome fragment's payload, parse it to model and complete the future with it
+            });
+        
+        return future;
+    }
+
+    //other methods we'll implement in a moment
+}
+```
+
+There's a couple of things to explain here. Firstly, our fetcher is abstract - it will have one abstract method for transforming payload into objects from our model (`Book`).
+Also, it's generic. That's because it doesn't matter for it what type will it return - it may be `Book`, it may be an array `Book[]` or it may be anything else.
+
+`TaskDataFetcher` implements `DataFetcher` (every data fetcher needs to) of type `CompletableFuture<T>`. Java implementation of GraphQL allows us to return futures instead of plain objects. If a data fetcher returns a future, graphQL will asyncronicly wait for it to complete. Only when all futures returned by data fetchers complete, the graphql will construct a response.
+
+Our abstract fetcher creates an instance of `FragmentsEngine`. The engine can execute task trees. Also, in the `get` method we create `FragmentEventContextTaskAware`. It's basicaly a task tree ready to be executed by the engine. Let's define the `setupTask` method that will take a configuration and parse it into a task tree. We'll also need one constant.

```

Comment:
```suggestion
Our abstract fetcher creates an instance of `FragmentsEngine`. The engine can execute task graphs. Also, in the `get` method we create `FragmentEventContextTaskAware`. It's basicaly a task graph ready to be executed by the engine. Let's define the `setupTask` method that will take a configuration and parse it into a task graph. We'll also need one constant.
```

Reply:
Trees are specific cases of graphs. The engine operates on task trees. "Graphs" could imply that cycles are possible.

Reply:
you're right but afaik we use the word `graph` everywhere else (in docs etc.)